### PR TITLE
ARROW-150 Improve documentation about limitations surrounding schemas and nested data structures

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/docs/source/schemas.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+PyMongoArrow Schemas and Their Usage
+====================================
+
+While there are many similarities between `MongoDB Projections <https://www.mongodb.com/docs/manual/reference/operator/projection/positional/>`_
+and :class:`pymongoarrow.api.Schema`'s, there are a few important differences:
+
+* Schemas do not support `"dotted" notation <https://www.mongodb.com/docs/manual/core/document/#dot-notation>`_,
+  instead they require that there can be only one type for any field name in a document **or any of its subdocuments**.
+* Sub documents must either follow the same Schema as their parent document, or must use field names that do not conflict
```
[the remark this answers is not in the document]
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: I fixed this based on new understanding.

File: bindings/python/docs/source/schemas.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   1  b'c\xec2\x98R(\xc9\x1e@#\xcc\xbc'         10      bar
+
+
+The same thing can also be accomplished using aggregation:
+
+.. code-block:: python
+
+   >>> from pymongo import MongoClient
+   >>> from pymongoarrow.api import Schema, find_pandas_all, aggregate_pandas_all
+   ...
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/docs/source/schemas.rst:121]: It wasn't clear to me that this example skipped lines. Why not remove the imports too in that case?

File: bindings/python/docs/source/schemas.rst [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+   >>>         "propStart": "$prop.start"
+   >>> },
+   >>> schema=Schema({"_id": ObjectIdType(), "propStart": int, "propName": str}))
+   >>> print(df)
+                                    _id  propStart propName
+   0  b'c\xec2\x98R(\xc9\x1e@#\xcc\xbb'          0      foo
+   1  b'c\xec2\x98R(\xc9\x1e@#\xcc\xbc'         10      bar
+
+
+The same thing can also be accomplished using aggregation:
```
Comment (ShaneHarvey) [inline bindings/python/docs/source/schemas.rst:115]: I suggest something like:
> For aggregate you can flatten the fields using the ``$project`` stage, like so:

File: bindings/python/docs/source/schemas.rst [hunk 4]
```diff
@@ -0,0 +1,138 @@
+.. _schema usage:
+
+PyMongoArrow Schemas and Their Usage
+====================================
+
+While there are many similarities between `MongoDB Projections <https://www.mongodb.com/docs/manual/reference/operator/projection/positional/>`_
```
Comment (ShaneHarvey) [inline bindings/python/docs/source/schemas.rst:6]: I think it's a little confusing to start off this page immediately talking about projections. I would expect a page titled "PyMongoArrow Schemas and Their Usage" to start off with basic examples and then expand into different use cases. Could we title this page `Schema Examples` instead? That way we can expand it in the future to show examples for arrays and other data types.

File: bindings/python/docs/source/schemas.rst [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+   >>>     "start": "string",
+   >>>     "prop": {
+   >>>         "name": "foo",
+   >>>         "start": 0,
+   >>>     }
+   >>>   }, {
+   >>>       "start": "string",
+   >>>       "prop": {
+   >>>         "name": "bar",
+   >>>         "start": 10,
```
Comment (ShaneHarvey) [inline bindings/python/docs/source/schemas.rst:34]: Can we collapse these documents onto one line to improve the readability?
```python
   >>> coll.insert_many([
   >>>     {"start": "string", "prop": {"name": "foo", "start": 0},
   >>>     {"start": "string", "prop": {"name": "bar", "start": 10}])
```

File: bindings/python/docs/source/schemas.rst [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+   >>>     }
+   >>>   },])
+   >>> df=find_arrow_all(coll,{}, schema=Schema({"start": str, "prop": struct([field("start", int32())])}))
```
Comment (ShaneHarvey) [inline bindings/python/docs/source/schemas.rst:37]: Please use consistent code style. Above we use `coll = MongoClient` and here we use `df=find_arrow_all(coll,{},...` without spaces.

File: bindings/python/docs/source/schemas.rst [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+   >>>         "start": 0,
+   >>>     }
+   >>>   }, {
+   >>>       "start": "string",
+   >>>       "prop": {
+   >>>         "name": "bar",
+   >>>         "start": 10,
+   >>>     }
+   >>>   },])
+   >>> df = find_pandas_all(coll, {}, schema=Schema({"start": str, "prop": struct([field("start", int32())])}))
```
Comment (ShaneHarvey) [inline bindings/python/docs/source/schemas.rst:71]: Let's remove the duplicate code and just show the find_pandas_all call.

File: bindings/python/docs/source/schemas.rst (hunk 6 above)
Comment (ShaneHarvey) [reply bindings/python/docs/source/schemas.rst:37]: `df` -> `arrow_table` which is what we call the find_arrow_all result elsewhere in the docs.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/docs/source/schemas.rst (hunk 6 above)
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: Done.

File: bindings/python/pymongoarrow/schema.py [hunk 8]
```diff
@@ -27,6 +27,8 @@ class Schema:
     Each key in ``schema`` is a field name and its corresponding value
     is the expected type of the data contained in the named field.
 
+    For more details about Schemas and their usage see :ref:`schema usage`.
```
Comment (ShaneHarvey) [inline bindings/python/pymongoarrow/schema.py:30]: This shows up in the docs as:
> For more details about Schemas and their usage see [PyMongoArrow Schemas and Their Usage](https://mongo-arrow--130.org.readthedocs.build/en/130/schemas.html#schema-usage).

How about:
> For more examples, see :ref:`schema usage`.

File: bindings/python/docs/source/schemas.rst (hunk 5 above)
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst:31]: Done.

File: bindings/python/docs/source/schemas.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst:107]: Done.

File: bindings/python/docs/source/schemas.rst (hunk 5 above)
Comment (blink1073) [reply bindings/python/docs/source/schemas.rst:31]: This bit is still not collapsed, was it reformatted by pre-commt?

Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: I think I changed the wrong part. Should be fixed now.

File: bindings/python/docs/source/schemas.rst [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+The following are a few examples of usage of PyMongoArrow Schemas in common situations.
+
+
+Nested Data With Schema
+-----------------------
+
+With aggregate or find methods, you can provide a schema for nested data using the struct object. Note that there can be conflicting
+names in sub-documents compared to their parent documents.
+
+.. code-block:: python
```
Comment (blink1073) [inline bindings/python/docs/source/schemas.rst:15]: I think we should adopt https://github.com/adamchainz/blacken-docs to clean up our docs.  

For this PR, that would mean:
- Change the code block type to `code-block:: pycon`
- Use `...` on continuation lines
- Get `blacken-docs` to parse
- Commit the changes

And then, open tickets for ARROW and PYTHON to add `blacken-docs` to our `pre-commit` hooks.

**Review by blink1073 — CHANGES_REQUESTED** [review]

File: bindings/python/docs/source/schemas.rst (hunk 7 above)
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: Done.

File: bindings/python/docs/source/schemas.rst (hunk 9 above)
Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: This makes sense to me. Done.

Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]:
```
(arrow3.10) ➜  docs git:(ARROW-150) ✗ blacken-docs source/schemas.rst
(arrow3.10) ➜  docs git:(ARROW-150) ✗ git add .
(arrow3.10) ➜  docs git:(ARROW-150) ✗ git commit -m "conform to blacken-docs"
```
`blacken-docs` exits without rewriting.

**Comment by blink1073** [comment]
The macos tests have started failing due to a lack of `pkg-config` on the box, I'll open a separate ticket for that.

File: bindings/python/docs/source/schemas.rst (hunk 9 above)
Comment (blink1073) [reply bindings/python/docs/source/schemas.rst]: Great!  Can you please open those two tickets?

Comment (juliusgeo) [reply bindings/python/docs/source/schemas.rst]: https://jira.mongodb.org/browse/ARROW-154 and https://jira.mongodb.org/browse/PYTHON-3610 !

**Comment by blink1073** [comment]
Can you please pull from `main` to pick up the macos fixes?

**Comment by juliusgeo** [comment]
@blink1073 Done.

**Review by blink1073 — APPROVED** [review]
LGTM!

**Pull request merged**